### DISCUSSION TITLE
Custom run loop

### DISCUSSION THREAD

**Discussion by helgoboss** [tokio-rs/tokio#2581] [General] [answered]
Hi, I've a question regarding the Tokio runtime. Is there any way to integrate it within an existing run loop instead of a firing up a new one?

Motivation: I would like to use some of Tokio's goodies (primarily timers) in a Rust plug-in for a C++ desktop application. This application has its own run loop which executes about 30x per second and which the plug-in can take part in. I wish there would be a way to repeatedly call something like `tokio_run_loop.idle()` in order to give Tokio a small time slice to drive futures one step further to completion. Using a runtime with `BasicScheduler` like this obviously doesn't work because it blocks until future completion.

**Comment by carllerche** [comment]
Could you elaborate on the event loop? Does it run 30x per second on a fixed timer? Is there an API to wake it up or control *when* it runs? How do you hook into the event loop?

**Reply by helgoboss** [reply]
Wow, that was fast, thanks.

That event loop is just the UI message loop of a typical desktop application. On Windows it's the Win32 message loop. "30x" is just an orientation. Depending on the stuff it does in each cycle, it can slow down or speed up. Not sure if it runs on a fixed timer.

There's no API to wake it up or control when it runs. It's continuously running. I hook into it by registering a callback function with the main application, which after that gets called on each cycle.

I'm not even sure if it makes sense to use Tokio for something like this, because it's definitely not a network application. But I'm interested in using some async/await code in my plug-in and work with delays and intervals - which Tokio offers. I could probably run the Tokio runtime in parallel with its own thread pool. But then I would always have to communicate with the thread that runs the main message loop whenever I want to interact with the application (since most of its functions must be called from the main thread only).

I hope that explanation makes sense.

**Reply by carllerche** [reply]
How does one do networking the "idiomatic" way (i.e. not w/ tokio) in this case? How do you send a message to the main loop?

**Reply by helgoboss** [reply]
Mostly I don't need to send a message at all because I'm already in the main loop (the host calls the plug-in in the main loop). If I'm in another thread, I can use a kind of queue, e.g. mpsc or crossbeam sender to send something to a receiver in the main loop. The receiver picks things up using `try_recv()`.

There's no idiomatic way for networking because it's not a networking application.

**Comment by Darksonn** [comment, marked as answer]
This is related to #2443.

**Reply by helgoboss** [reply]
You are right. This is it. I would need exactly the same.

**Comment by congzhangzh** [comment]
Maybe the same purpose?
https://github.com/tokio-rs/tokio/issues/6994
